### ISSUE TITLE
[Feature Request] Ability to refresh auth token

### ISSUE THREAD

**Issue by asarkar** [neo4j/neo4j-go-driver#429]
- Neo4j version: AWS Neptune
- Neo4j Mode: N/A
- Driver version: All
- Operating system: Linux

We're using openCypher with AWS Neptune and want to use the neo4j driver with the bolt protocol. We are generating an AWS V4 signature and passing it in as a basic auth to the driver. The request is signed using credentials that expires in 5 minutes.

Unfortunately, it seems that the authentication token is [cached at connection creation time](https://github.com/neo4j/neo4j-go-driver/blob/5.0/neo4j/auth_tokens.go#L24).  Thus, requests start failing after 5 minutes because the signature is no longer valid.

The Java driver provides a way to avoid this problem by exposing a [toMap](https://github.com/neo4j/neo4j-java-driver/blob/5.0/driver/src/main/java/org/neo4j/driver/internal/security/InternalAuthToken.java#L42) method, that is called to obtain the token. While the cached token is returned by default, this method can be overridden in a subclass as [shown here](https://docs.aws.amazon.com/neptune/latest/userguide/access-graph-opencypher-bolt.html#access-graph-opencypher-bolt-java-iam-auth).

We would like the Go driver to also provide an option not to cache the auth token and regenerate it on demand.

**Comment by fbiville** [comment]
This is already tracked internally and in other official driver repositories, as you are probably aware:

- https://github.com/neo4j/neo4j-python-driver/issues/834
- https://github.com/neo4j/neo4j-javascript-driver/issues/993
- https://github.com/neo4j/neo4j-dotnet-driver/issues/664

This is under consideration and may or may not be released in the coming months.
We will let you know.

**Comment by asarkar** [comment]
I'm not aware of your internal considerations, of course, but I'm aware of similar issues in the other driver repos. One of those is opened by me.

**Comment by xiazcy** [comment]
Hi @fbiville, just wondering is there an estimate of when this feature will be released for all the drivers? Thank you.

**Comment by AndyHeap-NeoTech** [comment]
We are currently aiming for the end of April. However this can change depending on how the work progresses.

**Comment by ghost** [comment]
@AndyHeap-NeoTech any updates? I have been kicking a couple of stories down the road waiting on this. How's it looking?

**Comment by AndyHeap-NeoTech** [comment]
Hi @firefox7025. I can confirm that the 5.8 release of the driver scheduled for tomorrow (2023/04/27) will contain the ability to refresh auth tokens (re-auth as we call it). It will also contain the related feature of user switching without creating a new driver instance.

Please note that this feature is considered to be in **Preview** for this release. The full definition of **Preview** can be found in the readme.

**Comment by fbiville** [comment]
Fixed by a9c87b9b66e844f4beab63b725725477fdbe6b83

**Comment by fbiville** [comment]
5.8.0 is out.
More details about re-auth here: https://github.com/neo4j/neo4j-go-driver/discussions/482
